Pump It Up

mjmagee991 opened the pull request:
Added simple heat pump driver
Created sample to demonstrate behavior and allow easy testing

ddp5730 committed:
Initialized Repository

ddp5730 committed:
Setup build and initial README.md

ddp5730 committed:
Set up auto-generating docs

ddp5730 committed:
Updated README.md

ddp5730 committed:
Update version variable

ddp5730 committed:
Updated hosting info in README.md and added LICENSE

ddp5730 committed:
Merge pull request #1 from RIT-EVT/feature/initial-template

Feature/initial template

cbolles committed:
Fix include order

light-orth committed:
Created initial code for sample pwm pump controller.

mjmagee991 committed:
Merge remote-tracking branch 'origin/feature/mjmagee991/requirements-documentation' into feature/light-orth/pump-control

# Conflicts:
#	targets/blinkUtil/main.cpp

mjmagee991 committed:
Fixed Formatting and Compilation Error

mjmagee991 committed:
Add GitHub Workflow and .clang-format

mjmagee991 committed:
Applied Clang-Format

mjmagee991 committed:
Merge pull request #2 from RIT-EVT/feature/mjmagee991/workflow-and-formatting

Workflow and Formatting

mjmagee991 committed:
Merge with Template

mjmagee991 committed:
Ignore cmake-build-debug

mjmagee991 committed:
Extracted Heat Pump Functionality to a Class

Also deleted example classes and target

mjmagee991 committed:
Modified Implementation to Better Match Requirements

Added stop() method for turning the pump off
Added macros to create constants
Changed speed values from a float (0-1) to a unit8_t (0-100) to reduce processing costs

mjmagee991 committed:
Added Documentation

cbolles commented on src/TMS/dev/HeatPump.cpp:
```diff
@@ -0,0 +1,36 @@
+
+namespace TMS {
+
+HeatPump::HeatPump(IO::PWM& pwm, uint8_t initSpeed) : pwm(pwm) {
```
For the `initSpeed` when the TMS system starts up, we do not want the pump running, so the pump should start out stopped.

mjmagee991 replied:
It's now been fully removed.

cbolles reviewed (changes requested):
Just need to make sure the pump can start up without running

ddp5730 commented on src/TMS/dev/HeatPump.cpp:
```diff
@@ -0,0 +1,36 @@
+
+HeatPump::HeatPump(IO::PWM& pwm, uint8_t initSpeed) : pwm(pwm) {
+    // Ensures initSpeed falls in the required range
+    if (initSpeed < MIN_INIT_SPEED) {
+        this->initSpeed = MIN_INIT_SPEED;
+    } else if (initSpeed > MAX_SPEED) {
+        this->initSpeed = MAX_SPEED;
+    } else {
+        this->initSpeed = initSpeed;
+    }
```
MIN_INIT_SPEED is 7.  But initSpeed is supposed to be ranged from 0-100.  Shouldn't this check if initSpeed < 0 here.

mjmagee991 replied:
The MIN_INIT_SPEED had to be 7 to make sure the duty cycle and period result in an impulse of >= 3 ms. I changed some things around and added documentation to make this clearer. Let me know if anything is still vague.

ddp5730 commented on src/TMS/dev/HeatPump.cpp:
```diff
@@ -0,0 +1,36 @@
+    pwm.setDutyCycle(initDutyCycle);
+}
+
+void HeatPump::setSpeed(uint8_t speed) {
+    // Ensures speed falls in the required range
+    if (speed > MAX_SPEED) {
+        speed = MAX_SPEED;
+    }
+
+    float dutyCycle = SPEED_TO_DUTY_CYCLE(speed);
+    pwm.setDutyCycle(dutyCycle);
+}
```
Do you apply the 3 ms pulse here when setting a speed after it was previously stopped?

mjmagee991 replied:
This is related to the minimum initialization speed. In the new implementation, the impulse is taken care of in this method. Let me know if that is clear now.

mjmagee991 committed:
Combined Functionality of init() into setSpeed() to make implementation clearer

Changed period to always be set to the same value
Removed init() and added isInit to implement functionality in setSpeed()
Added case in setSpeed() to deal with a speed of 0

Commit:
Applied Formatting Changes During GitHub Build

cbolles reviewed (approved):
Looks good!

ddp5730 reviewed (approved):
It makes more sense now and everything looks clean.  Nice work!
